BehaviorTreeEditor.cs

```diff
@@ -9,276 +9,391 @@
+					if ( !_behaviorTree.root )
+					{
+						Debug.Log( "Behavior tree root doesn't exist, creating a new one." );
+						_behaviorTree.root = CreateNodeWithoutParent( typeof( NullNode ) );
+					}
+
+					BuildEditorData();
+
+					CreateGUI();
+				}
+			}
+			catch ( Exception e )
+			{
+				Debug.LogError( e );
+			}
 		}
-		catch ( Exception e )
+
+		void CreateGUI()
 		{
-			Debug.LogError( e );
+			BeginWindows();
+			DrawNode( _behaviorTree.root );
+			EndWindows();
 		}
-	}
 
-	void CreateGUI()
-	{
-		CreateTypeLists();
-
-		_behaviorTree =
-		    (BehaviorTree)EditorGUILayout.ObjectField(
-		        "Behavior Tree",
-		        _behaviorTree,
-		        typeof( BehaviorTree ),
-		        false );
-
-		if ( _behaviorTree != null )
+		void DrawNode( TreeNode node )
 		{
-			// set default root node
-			if ( !_behaviorTree.root )
+			if ( node == null )
 			{
-				_behaviorTree.root = nullNode;
+				Debug.LogError( "Node is null!" );
+				return;
 			}
 
-			GUILayout.Label( "Nodes", EditorStyles.boldLabel );
+			EditorData nodeData = _editorData[node.id]; // TODO: Crash here after the editor resets.
 
-			Type resultType = CreateNodeTypeDropdown( _behaviorTree.root, 2 );
-			if ( resultType != _behaviorTree.root.GetType() )
+			// draw window
+			Rect resultRect = GUI.Window( nodeData.id, nodeData.nodeRect, DrawNodeWindow, DisplayName( node.ToString() ) );
+			if ( resultRect != nodeData.nodeRect )
 			{
-				DeleteNode( _behaviorTree.root );
-				_behaviorTree.root = CreateNode( resultType );
+				nodeData.nodeRect = resultRect;
+				EditorUtility.SetDirty( _behaviorTree );
 			}
+
+			// handle children
+			if ( node is Decorator )
+			{
+				Decorator decorator = (Decorator)node;
 			
-			CreateNodeGUI( _behaviorTree.root, 3 );
-		}
-	}
+				// ensure that the node has a child
+				if ( !decorator._child )
+				{
+					decorator._child = CreateNodeWithParent( typeof( NullNode ), decorator );
+				}
 
-	int CreateNodeGUI( TreeNode node, int height )
-	{
-		DebugUtils.Assert( node, "Cannot create GUI for null node!" );
+				EditorData childData = _editorData[decorator._child.id];
 
-		++_indentLevel;
-		
-		if ( node is Decorator )
-		{
-			height = CreateDecoratorGUI( (Decorator)node, height );
+				// handle drawing the child
+				DrawNodeCurve( nodeData.nodeRect, childData.nodeRect );
+				DrawNode( decorator._child );
+			}
+			else if ( node is Compositor )
+			{
+				Compositor compositor = (Compositor)node;
+
+				foreach ( TreeNode child in compositor._children )
+				{
+					EditorData childData = _editorData[child.id];
+
+					// handle drawing child
+					DrawNodeCurve( nodeData.nodeRect, childData.nodeRect );
+					DrawNode( child );
+				}
+			}
 		}
-		else if ( node is Compositor )
+
+		static string DisplayName( string name )
 		{
-			height = CreateCompositorGUI( (Compositor)node, height );
+			if ( name.Contains( '.' ) )
+			{
+				string[] tokens = name.Split( '.' );
+				return tokens[tokens.Length - 1];
```

Comment:
Since I put the nodes into the `BehaviorTree` their names default to stuff like `BehaviorTree.NullNode`. This splits the name string on the "." character and then takes the last token which should just be the node's name.

Now that I think about it, there might just be a way to get the unqualified name, so I'll look into that as an option.